### PR TITLE
[SPARK-44394][CONNECT][WEBUI] Add a Spark UI page for Spark Connect

### PR DESCRIPTION
**Pull request by jasonli-db**
## What changes were proposed in this pull request?
Add a new Spark UI page to display session and execution information for Spark Connect. This builds of the work in SPARK-43923 (https://github.com/apache/spark/pull/41443) that adds the relevant SparkListenerEvents and mirrors the ThriftServerPage in the Spark UI for JDBC/ODBC.

Screenshot 2023-07-27 at 11 29 22 PM

Screenshot 2023-07-27 at 11 29 15 PM

### Why are the changes needed?
This gives users a way to access session and execution information for Spark Connect via the UI and provides the frontend interface for the related SparkListenerEvents.

### Does this PR introduce _any_ user-facing change?
Yes, it will add a new tab/page in the Spark UI

### How was this patch tested?
Unit tests

### CODE REVIEW

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala
Comment (gengliangwang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala:124]: Why the default value is 200? For SQL query retention, the default value is 1000

**Comment by jasonli-db** [comment]
@juliuszsompolski @grundprinzip Can you help review this PR? This builds off of https://github.com/apache/spark/pull/41443, so the commit to look at is https://github.com/apache/spark/pull/41964/commits/fcf0ca69ea30a71e20ec15673d62bbfe91f11a36. 

Also cc @rednaxelafx @gengliangwang

**Comment by gengliangwang** [comment]
Since the Spark connect is more about SQL, shall we show the SQL execution links on the UI?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala:291]: make it `sparkContext: SparkContext` , not an option, and in SparkConnectServer pass it explcitly from the session there. better that than having to `.getOrElse(SparkSession.active.sparkContext)` without controlling if there's an active session here

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala
Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala:124]: This is what the Thrift Server has for defaults: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L1414. Can be changed though. Is 1000 preferred?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala:52]: wouldn't SparkConnectService.stop() have been called already somewhere and already detached it?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala:287]: nit: maybe move `createListenerAndUI` to `start`, just like `detach` is in `stop`. Then don't need to pass sc to `startGRPCService`

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala:186]: could both be private, or `private[connect]`?

**Review by juliuszsompolski — COMMENTED** [review]
reviewed connect service parts, didn't review `ui` package.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala
Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala]: Updated to `private[connect]`.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala
Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala]: Updated.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala
Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala:52]: Does it get called on this path? I only saw `SparkConnectService.server.awaitTermination()` on line 49. When I searched for SparkConnectService.stop(), I only saw two occurrences in [SimpleSparkConnectService](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/SimpleSparkConnectService.scala#L55) and [SparkConnectPlugin](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/SparkConnectPlugin.scala)

Comment (juliuszsompolski) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala:52]: yeah... it seems that in this path the Server is killed by a signal (via `sbin/stop-connect-server.sh`) and then `SparkConnectService.stop()` is never called.
It would be nice to SparkConnectService.stop() here, but then it will try to shutdown the server again... 
In any case, it's unrelated to this PR, so let's leave it like it is now.

**Comment by gengliangwang** [comment]
> Since the Spark connect is more about SQL, shall we show the SQL execution links on the UI?

@jasonli-db What is your opinion of this one?

**Comment by jasonli-db** [comment]
> > Since the Spark connect is more about SQL, shall we show the SQL execution links on the UI?
> 
> @jasonli-db What is your opinion of this one?
 I agree that adding the SQL execution links would be ideal and useful for the user. @juliuszsompolski Can you give your input on this and confirm whether or not this is currently possible. Do we have a link between SQL execution IDs and Connect requests?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala:76]: @jasonli-db could you try to use https://github.com/juliuszsompolski/apache-spark/commit/639d515d629d78825489510eb98b3648e4122b04
to make it
```
val jobTag = jobTags.findFirst({ case ExecuteJobTag(jobTag) => jobexeclistTag }
val exec = executionList.get(jobTag)
```

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerListener.scala:68]: Discussed offline with @jasonli-db and @jdesjean that for executions that don't start a Spark Job (e.g. SELECT 1), they will not be recorded. For that reason, as a followup, it would be best if jobTags could also be added to the SparkListenerSQLExecutionStart event.

**Review by juliuszsompolski — APPROVED** [review]
Reviewed and LGTM on the connect part; didn't review details of the UI and state store part.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala
Comment (gengliangwang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala:112]: Shall we use allow list to make it more robust?
E.g. check the state is among STARTED, COMPILED, READY

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerHistoryServerPlugin.scala
Comment (gengliangwang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerHistoryServerPlugin.scala:40]: increase the display order to avoid conflict with StreamingQueryHistoryServerPlugin?

Comment (rednaxelafx) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerHistoryServerPlugin.scala:40]: It'd be good to have this, but because of how other SHS plugins are also broken in the same way, it may be clearer to have a separate PR that fixes the displayOrder of all of the plugins in one go so that we can see the context of why the values were chosen in one single PR.

e.g. if we set this plugin's display order to `4`, it wouldn't be very obvious just from this PR alone why it should be a `4` while a lot of other plugins have a display order of `1`

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala
Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala]: Updated.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerHistoryServerPlugin.scala
Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerHistoryServerPlugin.scala:40]: Agreed we should update them all at once. Should we do it in a separate PR or just update all of them in this one?

Comment (rednaxelafx) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerHistoryServerPlugin.scala:40]: I prefer doing the wholesale update in a separate PR, but please check with the Apache Spark committers in this thread to confirm 🙏

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerPage.scala
Comment (gengliangwang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerPage.scala:475]: image
Is the ordering working?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerTab.scala
Comment (gengliangwang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerTab.scala:30]: nit: two spaces indent

**Comment by gengliangwang** [comment]
@jasonli-db Please try more tests. I did a quick test and find the SQL IDs need deduplication.
image

**Comment by gengliangwang** [comment]
Do we expect users to set the job tags? The generated tags are long and meaningless
image

**Comment by gengliangwang** [comment]
In the screenshot above, the close time and finish time column seems duplicated. And the execution time and duration also seems quite similar. Why do we have so many columns about the timings?

**Comment by gengliangwang** [comment]
Nit: we can have follow-up to improve the page when a session doesn't exist(e.g. evicted from KV store):
image

For example, id not existed is handled in the SQL page:
image

**Comment by jasonli-db** [comment]
> Do we expect users to set the job tags? The generated tags are long and meaningless image

@juliuszsompolski @jdesjean Is there any use in including the job tags? Actually now that I think about it, since it's constructed from the SessionID, userID, and operationID, I think we can instead just add a column for the operationID and remove the jobTag column. What do you think?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerTab.scala
Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerTab.scala:30]: This seems to be set by the linter (it reverted back to this form after I ran the lint command).

**Comment by juliuszsompolski** [comment]
OperationId could indeed maybe be more useful instead. If a user uses connect api like SparkSession.interruptOperation, they pass just operationId.
The only time a user would want to use jobTag is if they wanted to directly call SparkContext.interruptJobsWithTag, which is a rather internal use.
We could consider displaying SparkListenerConnectOperationStarted.sparkSessionTags, because these are uses settable (via connect SparkSession.addTag / removeTag / clearTags), and they are used by users to call connect SparkSession.interruptTag.

**Comment by juliuszsompolski** [comment]
> In the screenshot above, the close time and finish time column seems duplicated. And the execution time and duration also seems quite similar. Why do we have so many columns about the timings?

@gengliangwang Finish time is finish execution, close time is when all the results were transferred; execution time is the time query was executing, while duration is end to end. It is meaningful for queries with large results, or e.g. complex queries that spend a lot of time in optimizer. It's the same with Thriftserver UI.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerHistoryServerPlugin.scala
Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerHistoryServerPlugin.scala]: Hm, I checked and since this is only incorrect in one other place (StreamingQueryHistoryServerPlugin.scala), I just updated it in this PR

**Comment by jasonli-db** [comment]
> @jasonli-db Please try more tests. I did a quick test and find the SQL IDs need deduplication. image

Thanks for catching this. It's been fixed

**Comment by jasonli-db** [comment]
> Nit: we can have follow-up to improve the page when a session doesn't exist(e.g. evicted from KV store): image
> 
> For example, id not existed is handled in the SQL page: image

Addressed to match the behavior for the SQL page.

**Comment by jasonli-db** [comment]
> OperationId could indeed maybe be more useful instead. If a user uses connect api like SparkSession.interruptOperation, they pass just operationId. The only time a user would want to use jobTag is if they wanted to directly call SparkContext.interruptJobsWithTag, which is a rather internal use. We could consider displaying SparkListenerConnectOperationStarted.sparkSessionTags, because these are uses settable (via connect SparkSession.addTag / removeTag / clearTags), and they are used by users to call connect SparkSession.interruptTag.

@juliuszsompolski @gengliangwang I've updated the screenshots. I added the operationID and sparkSessionTags. For now, I kept the jobTag since it seems it could still be used if only internally. I'll remove it if/when we come to a final consensus to do so. Please take a look!

**Comment by gengliangwang** [comment]
@jasonli-db I have to say to the current table is quite wide. 
image

Can we at least put the operationID column behind the column "State" since it is not commonly useful?


A good way to resolve this is to have checkboxes and make operationID/sparkSessionTags/jobTag as  additional metrics:
image
This can be done in a followup

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala:131]: For potential followup:
I find it a bit weird, that it doesn't have a RUNNING execution state, but it's actually the same as in Thriftserver:
https://github.com/apache/spark/blob/master/sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala#L119C58-L119C58

Spark Connect records that it's "ready for execution" at the end of planner (or before eager commands), and we record "finished" when the result stage completes... Should we have another event "RUNNING" that e.g. for SparkConnectPlanExecution would be right before `spark.sparkContext.submitJob`, and for Commands would need to be determined on a per command basis. WDYT @jdesjean ; that would be different from Thriftserver, but it seems to me like that "RUNNING" state is missing from Thriftserver... cc @wangyum @yaooqinn  @bogdanghit WDYT from the Thriftserver perspective?

Comment (bogdanghit) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala:131]: I think a `RUNNING` state should cover the period between it enters `STARTED` state until it enters one of the terminal states. Using this definition, in ThriftServer we can infer the execution is in running / active state if it's not `FAILED, CANCELED, TIMEDOUT or CLOSED.`

Nit: It would be good to keep these states consistent. The SparkConnect execution state has a `READY` state which is missing in ThriftServer.

Comment (juliuszsompolski) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala:131]: @bogdanghit see https://github.com/apache/spark/pull/41748 from @jdesjean . Jean did some work to be able to figure out the "READY" state to come before the start of eager commands. In Thriftserver there's a bit of ugly issue that eager commands would execute even before the COMPILED state, because they are eagerly executed straight in analysis.
Thriftserver could use https://github.com/apache/spark/pull/41748 to also display information about eager execution better.

Comment (jasonli-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerAppStatusStore.scala:131]: I have the UI display the state as RUNNING is isActiveExecution is true, which covers STARTED, COMPILED, READY: https://github.com/apache/spark/pull/41964/commits/8abdecdc7cedb6ff5a13090dd4bcae2ecf50d0d0#diff-ef2b5b937c2c5e83d6ca9804c9d402d263bf802550cc8daeead089498c123aa0L322.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerPage.scala
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ui/SparkConnectServerPage.scala:322]: LGTM, I think it makes more sense, thanks!

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
@jasonli-db Thanks for the work. I will merge the PR once the tests are passed.

**Comment by gengliangwang** [comment]
Thanks, merging to master/3.5

**Comment by gengliangwang** [comment]
@jasonli-db This PR has conflict with the branch-3.5
Please cherry-pick it before the RC1 on Aug 1

**Comment by juliuszsompolski** [comment]
Thanks @gengliangwang . Since the conflict was trivial (imports), I resolved and raised https://github.com/apache/spark/pull/42224 for branch-3.5.

**Pull request closed without merging**
